File: src/vct/col/ast/ClassType.java
package vct.col.ast;

import static hre.System.Abort;
import static hre.System.Debug;
import static hre.System.Warning;

import java.util.Arrays;

public class ClassType extends Type {

  private final String name[];
  private final ASTNode args[];
  
  public ClassType(String name[],ASTNode ... args){
    this.name=Arrays.copyOf(name,name.length);
    this.args=Arrays.copyOf(args,args.length);
  }
  
  public ASTNode[] getArgs(){
    return args;
  }
  
  public ASTNode getArg(int i){
    return args[i];
  }


  public String[] getNameFull(){
    return Arrays.copyOf(name,name.length);
  }
  
  public ClassType(String name) {
    this.name=new String[1];
    this.name[0]=name;
    args=null;
  }
  public ClassType(String name[]) {
    if (name.length==0) Abort("illegal name of length 0");
    this.name=Arrays.copyOf(name,name.length);
    args=null;
  }
  public String getName(){
    return name[name.length-1];
  }
  public String getFullName(){
    String res=name[0];
    for(int i=1;i<name.length;i++){
      res+="."+name[i];
    }
    return res;
  }

  public void accept_simple(ASTVisitor visitor){
    visitor.visit(this);
  }

  public boolean supertypeof(ProgramUnit context, Type t){
    if (t instanceof ClassType) {
      ClassType ct=(ClassType)t;
      // java.lang.Object is supertype of everything.
      if (name.length==3 && name[0].equals("java") &&
          name[1].equals("lang") && name[2].equals("Object")) return true;
      // Everything is a supertype of <<null>>.
      if (ct.name.length==1 && ct.name[0].equals("<<null>>")) return true;
      if (ct.name.length==1 && ct.name[0].equals("<<label>>")) return true;
      return supertype_search(context,ct);
    }
    return false;
  }
  
  
  private boolean supertype_search(ProgramUnit context, ClassType ct) {
    Debug("checking if %s is asuper type of %s",this,ct);
    if (ct.name.length==name.length){
      int i=0;
      while(i<name.length){
        if (!ct.name[i].equals(name[i])) break;
        i++;
      }
      if (i==name.length) return true;
    }
    Debug("argument is not the same type not the same type");
    if (context==null) {
      Debug("missing context");
      return false;
    }
    ASTClass cl=context.find(ct);
    if (cl==null) {
      Debug("class not found");
      return false;
    }
    for(ClassType ct_parent:cl.super_classes){
      if (supertype_search(context,ct_parent)) return true;
    }
    return false;
  }

  public String toString(){
    return getFullName();
  }
  
  public int hashCode(){
    return getFullName().hashCode();
  }
  public boolean equals(Object o){
    if (o instanceof ClassType) {
      return getFullName().equals(((ClassType)o).getFullName());
    } else {
      return false;
    }
  }
  public ASTNode zero(){
    return new NameExpression(NameExpression.Kind.Reserved,"null");
  }
  
  public boolean isNull(){
    return name.length==1 && name[0].equals("<<null>>"); 
  }
}

File: vercors/src/main/java/vct/col/ast/ClassType.scala
package vct.col.ast

import hre.lang.System.Debug
import vct.col.util.VisitorHelper

object ClassType {
  val nullType = new ClassType("<<null>>")
  val labelType = new ClassType("<<label>>")
  
  /** Tests whether `name` holds the class type `java.lang.Object`. */
  def isJavaLangObject(name:List[String]) : Boolean =
    name == List("java","lang","Object") || name == List("java_DOT_lang_DOT_Object")
  
  /** Tests whether `name` holds a default type name, for example "null" or "label". */
  def isDefaultName(name:List[String]) : Boolean = 
    name == nullType.names || name == labelType.names
}

/**
 * AST node that represents the type of classes (including their class parameters).
 * 
 * @param names A list of name parts that together constitute the full class name (including package name)
 * @param params A list of AST nodes representing the types of the class parameters
 * @author sccblom, whmoortwijn
 */
case class ClassType(val names:List[String], val params:List[ASTNode]) extends Type(params) with VisitorHelper {
  require(!names.isEmpty, "class types must have a name (at least one name part).")
  
  var definition : ASTDeclaration = null
  
  def this(names:Array[String], args:Array[ASTNode]) = this(names.toList, args.toList)
  def this(names:Array[String]) = this(names, Array[ASTNode]())
  def this(name:String) = this(Array(name))
  
  def getName = this.names.last
  def getNameFull = this.names.toArray
  def getFullName(separator:String) = this.names mkString separator
  def getFullName : String = getFullName(".")
  def setDefinition(decl:ASTDeclaration) = definition = decl

  /**
   * Checks if any of `foundClass`'s super classes (of implemented classes) is a supertype of `this`,
   * in the given program context (`context`).
   */
  private def searchContextForSupertype(unit:ProgramUnit, foundClass:Option[ASTClass]) = foundClass match {
    case None => Debug("class not found."); false
    case Some(cl:ASTClass) => {
      cl.super_classes.exists { parent => searchForSupertype(Some(unit), parent) } ||
      cl.implemented_classes.exists { parent => searchForSupertype(Some(unit), parent) }
    }
  }
  
  /**
   * Checks if the type of this object (i.e. `this`) is a supertype of `ct` in the 
   * given program context (`context`).
   */
  private def searchForSupertype(context:Option[ProgramUnit], ct:ClassType) : Boolean = {
    Debug(s"checking if $this is a supertype of $ct.")
    if (ct.names == this.names) true else context match {
      case Some(unit:ProgramUnit) => searchContextForSupertype(unit, Option(unit.find(ct)))
      case None => Debug("missing program context."); false
    }
  }
  
  /**
   * @note Everything is a supertype of `null` (or any other default name).
   */
  private def searchForSupertype(context:ProgramUnit, other:Type) : Boolean = other match {
    case ct:ClassType => ClassType.isDefaultName(ct.names) || searchForSupertype(Option(context), ct)
    case _ => false
  }
  
  /**
   * Tests whether this class is a supertype of `otherType` (in the given `context`).
   * 
   * @note `java.lang.Object` is a supertype of everything.
   */
  override def supertypeof(context:ProgramUnit, otherType:Type) = 
    ClassType.isJavaLangObject(this.names) || searchForSupertype(context, otherType)
  
  override def equals(obj:Any) = obj match {
    case other:ClassType => this.getFullName == other.getFullName
    case _ => false
  }
  
  override def hashCode = this.getFullName.hashCode
  override def toString = this.getFullName + super.toString
  override def isNull = this.names == ClassType.nullType.names
  override def zero = new NameExpression(ASTReserved.Null)
  
  override def accept_simple[T,A](m:ASTMapping1[T,A], arg:A) = m.map(this, arg)
  override def accept_simple[T](v:ASTVisitor[T]) = handle_standard(() => v.visit(this))
  override def accept_simple[T](m:ASTMapping[T]) = handle_standard(() => m.map(this))
  override def accept_simple[T](m:TypeMapping[T]) = handle_standard(() => m.map(this))
}
